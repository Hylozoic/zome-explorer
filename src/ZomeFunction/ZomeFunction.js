import React, { useState } from 'react'
import ExpandButton from '../ExpandButton'
import CallRecord from '../CallRecord'

export default function ZomeFunction ({ fnDeclaration, callZomeFunc, zomeName }) {
  const defaultParams = 
    `{\n${fnDeclaration.inputs.reduce((acc, { name }) => acc + `  "${name}": "",\n`, '')}}`
  const [params, setParams] = useState(defaultParams)
  const { name } = fnDeclaration  

  const [callRecord, setCallRecord] = useState()
  const call = async () => setCallRecord(await callZomeFunc(name)(JSON.parse(params)))

  const [expanded, setExpanded] = useState(false)
  const toggleExpanded = () => setExpanded(!expanded)


  return <div className='zome-function'>
    <div className='zome-function-header clickable-div' onClick={toggleExpanded}>
      <div className='function-name'>
        {name}
        <ExpandButton expanded={expanded} toggleExpanded={toggleExpanded} />
      </div>
      {expanded && <div className='zome-function-signature'>
        <div className='zome-function-signature-label'>Inputs</div>
        {fnDeclaration.inputs.map(({ name, type }) => <div className='zome-function-signature-row' key={name}>
          {name}: {type}
        </div>)}
        <div className='zome-function-signature-label'>Outputs</div>
        {fnDeclaration.outputs.map(({ name, type }) => <div className='zome-function-signature-row' key={name}>
          {name}: {type}
        </div>)}
      </div>}
    </div>
    {expanded && <TabableEditor id={`editor-${zomeName}-${name}`} value={params} onChange={({ target: { value } }) => setParams(value)} />}
    {expanded && <button className='call-button' onClick={call}>Call</button>}
    {callRecord && <CallRecord callRecord={callRecord} />}
  </div>
}

export function TabableEditor ({ id, value, onChange }) {
  const onKeyDown = e => {
    if (e.keyCode === 9) {
      e.preventDefault()      
      insertAtCaret(id, '  ')
    }
  }

  return <textarea
    id={id}
    rows={15} 
    cols={50} 
    value={value}
    onChange={onChange} 
    onKeyDown={onKeyDown} />
}

function insertAtCaret(areaId, text) {
  var txtarea = document.getElementById(areaId);
  var scrollPos = txtarea.scrollTop;
  var caretPos = txtarea.selectionStart;

  var front = (txtarea.value).substring(0, caretPos);
  var back = (txtarea.value).substring(txtarea.selectionEnd, txtarea.value.length);
  txtarea.value = front + text + back;
  caretPos = caretPos + text.length;
  txtarea.selectionStart = caretPos;
  txtarea.selectionEnd = caretPos;
  txtarea.focus();
  txtarea.scrollTop = scrollPos;
}