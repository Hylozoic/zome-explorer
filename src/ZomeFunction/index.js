import ZomeFunction from './ZomeFunction'
export default ZomeFunction