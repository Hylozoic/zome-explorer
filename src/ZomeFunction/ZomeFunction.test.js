import ZomeFunction, { getDefaultParams, TabableEditor } from './ZomeFunction'
import { shallow } from 'enzyme'
import React from 'react'

const fnDeclaration = {
  name: 'the function',
  inputs: [
    {
      name: 'input-one',
      type: 'String'
    },
    {
      name: 'input-two',
      type: 'bool'
    },
    {
      name: 'input-three',
      type: 'Vec<String>'
    },
  ]
}

describe('getDefaultParams', () => {
  expect(getDefaultParams(fnDeclaration)).toMatchSnapshot()
})


describe('ZomeFunction', () => {
  it('matches the last snapshot', () => {
    const props = {
      fnDeclaration,
      callZomeFunc: name => () => {},
      zomeName: 'the zome'
    }  
    const wrapper = shallow(<ZomeFunction {...props} />)
    expect(wrapper).toMatchSnapshot()
  })  
})

describe('TabableEditor', () => {
  it('matches the last snapshot', () => {
    const props = {
      id: 'id',
      value: 'text',
      onChange: () => {}
    }  
    const wrapper = shallow(<TabableEditor {...props} />)
    expect(wrapper).toMatchSnapshot()
  })  
})