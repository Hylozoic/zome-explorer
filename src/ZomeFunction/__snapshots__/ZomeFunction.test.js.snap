// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"{
  \\"input-one\\": \\"\\",
  \\"input-two\\": false,
  \\"input-three\\": [],
  \\"input-four\\": { },
  \\"input-five\\": \\"\\"
}"
`;

exports[`TabableEditor matches the last snapshot 1`] = `
<textarea
  cols={50}
  id="id"
  onChange={[Function]}
  onKeyDown={[Function]}
  rows={15}
  value="text"
/>
`;

exports[`ZomeFunction matches the last snapshot 1`] = `
<div
  className="zome-function"
>
  <div
    className="zome-function-header clickable-div"
    onClick={[Function]}
  >
    <div
      className="function-name"
    >
      the function
      <ExpandButton
        expanded={false}
        toggleExpanded={[Function]}
      />
    </div>
  </div>
</div>
`;
